projects/plugins/boost/app/assets/src/js/features/critical-css/local-generator/local-generator-provider.tsx

```diff
@@ -0,0 +1,129 @@
+	setProviderProgress: ( progress: number ) => void;
+};
+
+type ProviderProps = {
+	children: ReactNode;
+};
+
+const CssGeneratorContext = createContext< LocalGeneratorContext | null >( null );
+
+/**
+ * Local Critical CSS Context Provider component - provides context for any descendants that want to
+ * either initiate the local Critical CSS generator, or check its status.
+ *
+ * @param {ProviderProps} props - Component props.
+ */
+export default function LocalCriticalCssGeneratorProvider( { children }: ProviderProps ) {
+	const [ abortController, setAbortController ] = useState< AbortController | undefined >(
+		undefined
+	);
+	const [ providerProgress, setProviderProgress ] = useState< number | undefined >( undefined );
+
+	const value = {
+		abortController,
+		setAbortController,
+		providerProgress,
+		setProviderProgress,
+	};
+
+	return <CssGeneratorContext.Provider value={ value }>{ children }</CssGeneratorContext.Provider>;
+}
+
+/**
+ * Internal helper function: Use the raw Critical CSS Generator context, and verify it's inside a provider.
+ */
+function useLocalCriticalCssGeneratorContext() {
+	const status = useContext( CssGeneratorContext );
+
+	if ( ! status ) {
+		throw new Error( 'Local critical CSS generator status not available' );
+	}
+
+	return status;
+}
+
+/**
+ * For status consumers: Get an overview of the local critical CSS generator status. Is it running or not?
+ */
+export function useLocalCriticalCssGeneratorStatus() {
+	const status = useLocalCriticalCssGeneratorContext();
+
+	return {
+		isRunning: !! status.abortController,
+	};
+}
+
+/**
+ * For Critical CSS UI: Actually run the local generator and return its status.
+ */
+export function useLocalCriticalCssGenerator() {
+	// Local Generator status context.
+	const { abortController, setAbortController, providerProgress, setProviderProgress } =
+		useLocalCriticalCssGeneratorContext();
+
+	// Critical CSS state and actions.
+	const [ cssState, setCssState ] = useCriticalCssState();
+	const setProviderCssAction = useSetProviderCssAction();
+	const setProviderErrorsAction = useSetProviderErrorsAction();
+	const generateCriticalCssAction = useRegenerateCriticalCssAction();
+
+	// Proxy nonce - reqiured config for the generator.
+	const proxyNonce = useProxyNonce();
+
+	useEffect(
+		() => {
+			if ( cssState.status === 'pending' && ! abortController ) {
+				// Start the local generator if the css state is pending.
+				setAbortController(
+					runLocalGenerator( cssState.providers, proxyNonce, {
+						onError: ( error: Error ) => setCssState( criticalCssErrorState( error.message ) ),
+						onFinished: () => setAbortController( undefined ),
+						setProviderCss: ( key: string, css: string ) =>
+							setProviderCssAction.mutateAsync( { key, css } ),
+						setProviderErrors: ( key: string, errors: CriticalCssErrorDetails[] ) =>
+							setProviderErrorsAction.mutateAsync( { key, errors } ),
+						setProviderProgress,
+					} )
+				);
```

Comment:
I'm not sure I understand why other effect dependencies are excluded. Maybe more comments would help? Or maybe there's a way to design the effect so that we don't need to break the rules?

Reply:
In fa2a998122, I added a comment explaining why. Basically this effect controls a process that is heavy to start and stop. If it depends on something like `cssState` itself, then every time any change is made to `cssState` it will stop the generator, then re-run the effect to figure out if it should start again.

This should only run when the actual status string changes. :)